README.md
# README
# Cryptbin

> A front end encrypted pastebin

This repository is a server designed to maintain a front end encrypted pastebin, a. la. `0bin.net`.

## Installing

See `doc/INSTALL.txt`.

## Contributing

See `doc/CONTRIBUTING.md`

## Licence.

This is licensed under Apache 2. See `doc/LICENCE.txt`.


doc/CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).

## [0.2.0] - 2018-02-04
### Added
- Frontend encryption 

## [0.1.0] - 2018-02-04
### Added
- Added AUTHORS.md
- Added CHANGELOG.md
- Added CONTRIBUTING.md
- Added INSTALL.md
- Added LICENCE.txt


main.go
// Cryptbin
// (C) Copyright George Saussy 2018
// version 0.1
package main

import (
  "database/sql"
  "flag"
  "fmt"
  "html/template"
  "net/http"
  "time"
  _ "github.com/lib/pq"
  _ "github.com/mattn/go-sqlite3"
)

// Structs

// ServerConfig contains the configuration of the server. 
// Should contain all of the global information for the server.
type ServerConfig struct {
  InProduction bool
  Debug bool
  Database *sql.DB
}

// Message struct is designed to hold a stored message. 
// The struct can also be passed to the view template to render
// the message.
type Message struct {
  Key string
  Text string
  TimeLeftMsg string
}

// Global variables
var config *ServerConfig

// Functions

// getBlankConfig returns a pointer to a blank ServerConfig object
func getBlankConfig() *ServerConfig {
  ret :=&ServerConfig{InProduction: false, Debug: false, Database: nil}
  return ret
}

// pasteHandler responds to the `/paste/` URI and sends the
// paste home page located at `html/paste.html`.
func pasteHandler(w http.ResponseWriter, r *http.Request) {
  logInfo("Rendering paste.")
  t, _ := template.ParseFiles("html/paste.html")
  t.Execute(w, nil)
}

// newHandler responds to POST requests posting a new form.
// A correctly formated POST is then redirected to the view 
// the submited content.
func newHandler(w http.ResponseWriter, r *http.Request) {
  logInfo("Computing new.")
  created_at := time.Now()
  paste_value := r.FormValue("paste_value")
  expiration_text := r.FormValue("expiration")
  expiration := getExpirationTime(expiration_text)
  burn_after_reading := (expiration_text == "burn_after_reading")
  paste_hash := hashString(paste_value)
  paste_key := r.FormValue("paste_key")
  logDebug("Value: "+paste_value+"\nExpiration: "+expiration_text+"\nHash: "+paste_hash+"\nKey: "+paste_key)
  storeNewPaste(created_at, paste_value, expiration, burn_after_reading, paste_hash)
  http.Redirect(w, r, "/view/"+paste_hash+"#"+paste_key, 301)
}

// aboutHandler responds to the `/about/` URI and sends the
// about page for Cryptbin located at `html/about.html`.
func aboutHandler(w http.ResponseWriter, r *http.Request) {
  logInfo("Rendering about.")
  t, _ := template.ParseFiles("html/about.html")
  t.Execute(w, nil)
}

// viewHandler responds to the `/view/` URI and sends the 
// view page template located at `html/view.html` using
// information given in te URI.
func viewHandler(w http.ResponseWriter, r *http.Request) {
  logInfo("Rendering view.")
  paste_hash := r.URL.Path[len("/view/"):]
  if checkValidHash(paste_hash) {
    paste_value, err := queryForPasteValue(paste_hash)
    time_left_msg := getTimeLeftMsg(paste_hash)
    if err != nil {
      _404Handler(w, r)
    } else {
      logDebug("hash: "+paste_hash+"\npaste: "+paste_value)
      cleanDatabase()
      dat := &Message{Key: paste_hash, Text: paste_value, TimeLeftMsg: time_left_msg}
      t, _ := template.ParseFiles("html/view.html")
      t.Execute(w, dat)
    }
  } else {
    _404Handler(w, r)
  }
}

// _404Handler routes the user to a 404 page
func _404Handler(w http.ResponseWriter, r *http.Request) {
  logInfo("Rendering 404.")
  t, _ := template.ParseFiles("html/404.html")
  t.Execute(w, nil)
}

// makeHandler wraps the HandlerFuncs to pass the server configuration.
func makeHandler(fn func(http.ResponseWriter, *http.Request)) http.HandlerFunc {
  return func(w http.ResponseWriter, r *http.Request) {
    fn(w, r)
  }
}

// main function is just the main function of this code base.
// As of right now, all that main does is 
// parse the command line arguments,
// initialize global variables, and 
// initialize an http server accordingly. 
func main() {
  // Parse flags
  in_production :=  flag.Bool("prod", false, "Set as true if the server is running in production.")
  debug := flag.Bool("debug", false, "Set to true to print debugging information to log.")
  db_password := flag.String("pwd", "", "Set the Postgres database password")
  db_username := flag.String("uname", "", "Set the Postgres username.")
  db_name := flag.String("dbname", "", "Set the Postgres database name.")
  flag.Parse()
  config = &ServerConfig{InProduction: *in_production, Debug: *debug, Database: nil}
  // Init database if need be
  if * in_production {
    dbinfo := fmt.Sprintf("user=%s password=%s dbname=%s sslmode=disable", *db_username, *db_password, *db_name)
    fmt.Printf(dbinfo)
    db_instance, err := sql.Open("postgres", dbinfo)
    if err != nil {
      logError(err.Error())
    }
    config.Database = db_instance
  } else {
    config.Database, _ = sql.Open("sqlite3", "db/cryptbin.db")
  }
  // Set up http server
  http.HandleFunc("/paste/", makeHandler(pasteHandler)) // Route `paste` to ''
  http.HandleFunc("/new/", makeHandler(newHandler)) // Route `new` to ''
  http.HandleFunc("/view/", makeHandler(viewHandler)) // Route `view` to ''
  http.HandleFunc("/about/", makeHandler(aboutHandler)) // Route `about` to ''
  http.HandleFunc("/", makeHandler(pasteHandler)) // Route root to `paste`
  http.Handle("/assets/", http.StripPrefix("/assets/", http.FileServer(http.Dir("html/assets")))) // Route `assets/*` to `html/assets` 
  if * in_production {
    fmt.Printf("Running Cryptbin in production mode\n")
    fmt.Printf("Go to localhost:8080\n")
    err := http.ListenAndServe(":8080", nil)
    fmt.Printf(err.Error())
    fmt.Printf("\n")
  } else {
    fmt.Printf("Running Cryptbin in developer mode\n")
    fmt.Printf("Go to localhost:8000\n")
    err := http.ListenAndServe(":8000", nil)
    fmt.Printf(err.Error())
    fmt.Printf("\n")
  }
}


query.go
package main

import (
 "hash/fnv"
 "strconv"
 "strings"
 "time"
 "errors"
)

// checkAndInitDataBase checks if a database is present.
// If no database present, it initializes one.
func checkAndInitDataBase() {
  if config.Database != nil {
    logDebug("Database initialized!")
  }
}

// hashString wraps the FNV hash function to hash strings
func hashString(s string) string {
  ret := ""
  h := fnv.New64a()
  h.Write([]byte(s))
  ret = strconv.FormatUint(h.Sum64(), 16)
  return ret
}

// checkValidHash checks if a string is a hexidecimal string
func checkValidHash(paste_hash string) bool {
  ret := true
  valid_chars := "0123456789abcdefABCDEF"
  for ind := 0; ind < len(paste_hash) && ret; ind++ {
    if ! strings.Contains(valid_chars, paste_hash[ind:ind+1]) {
      ret = false
    }
  }
  return ret
}

// getExpirationTime get expiration time from the POST parameter
func getExpirationTime(expiration_text string) time.Time {
  ret := time.Now()
  switch expiration_text {
  case "burn_after_reading":
    ret = time.Now().Add(time.Hour*12)
  case "12_hr":
    ret = time.Now().Add(time.Hour*12)
  case "24_hr":
    ret = time.Now().Add(time.Hour*24)
  case "5_days":
    ret = time.Now().Add(time.Hour*5*24)
  default:
    ret = time.Now().Add(time.Hour*12) // Default to 12 hr 
  }
  return ret
}

// storeNewPaste stores a paste in the database 
func storeNewPaste(created_at time.Time, paste_value string, expiration time.Time, burn_after_reading bool, paste_hash string) {
  if config.InProduction {
    var lastInsertId string
    err := config.Database.QueryRow("INSERT INTO paste(created_at, paste_value, expiration, burn_after_reading, paste_hash, view_count) VALUES($1, $2, $3, $4, $5, $6) returning paste_hash;", created_at, paste_value, expiration, burn_after_reading, paste_hash, 0).Scan(&lastInsertId)
    if err != nil {
      logError(err.Error())
    }
  } else {
    stmt, err := config.Database.Prepare("INSERT INTO paste(created_at, paste_value, expiration, burn_after_reading, paste_hash, view_count) values(?, ?, ?, ?, ?, ?);")
    if err != nil {
      logDebug("Error 1")
      logError(err.Error())
    }
    _, err = stmt.Exec(created_at, paste_value, expiration, burn_after_reading, paste_hash, 0)
    if err != nil {
      logDebug("Error 2")
      logError(err.Error())
    }
  }
}

// queryForPasteValue runs a query on the database to get the paste value associated by the hash
func queryForPasteValue(paste_hash string) (string, error) {
  ret := ""
  rows, err := config.Database.Query("SELECT paste_value FROM paste WHERE paste_hash='" + paste_hash + "'")
  for rows.Next() {
    rows.Scan(&ret)
  }
  rows.Close()
  config.Database.Exec("UPDATE paste SET view_count = view_count + 1 WHERE paste_hash='" + paste_hash + "'")
  if ret == "" || err != nil {
    err = errors.New("Hash not in database.")
  }
  return ret, err
}

// cleanDatabase deletes all rows from the database that are too old
func cleanDatabase() {
  if config.InProduction {
    stmt, err := config.Database.Prepare("DELETE FROM paste WHERE view_count > 1 AND burn_after_reading")
    if err != nil {
      logError("Error 1")
      logError(err.Error())
    }
    _, err = stmt. Exec()
    if err != nil {
      logError("Error 2")
      logError(err.Error())
    }
    stmt, err = config.Database.Prepare("DELETE FROM paste WHERE expiration < NOW()")
    if err != nil {
      logError("Error 3")
      logError(err.Error())
    }
    _, err = stmt.Exec()
    if err != nil {
      logError("Error 4")
      logError(err.Error())
    }
  } else {
    config.Database.Exec("DELETE FROM paste WHERE expiration < DATETIME('now')")
    config.Database.Exec("DELETE FROM paste WHERE view_count > 1 AND burn_after_reading = 1")
  }
}

// checkIfBurnAfterReading checks if a paste is set to burn after reading
func checkIfBurnAfterReading(paste_hash string) bool {
  ret := false
  rows, _ := config.Database.Query("SELECT burn_after_reading FROM paste WHERE paste_hash='" + paste_hash + "'")
  for rows.Next() {
    rows.Scan(&ret)
  }
  rows.Close()
  return ret
}

// getExpiration gets the time.Time at which a paste is expected to expire.
func getExpiration(paste_hash string) time.Time {
  ret := time.Now()
  rows, _ := config.Database.Query("SELECT expiration FROM paste WHERE paste_hash='" + paste_hash + "'")
  for rows.Next() {
    rows.Scan(&ret)
  }
  rows.Close()
  return ret
}

// getTimeLeftMsg get the message to alert users how long a message
// has left to live
func getTimeLeftMsg(paste_hash string) string {
  ret := ""
  burn_after_reading := checkIfBurnAfterReading(paste_hash)
  expiration := getExpiration(paste_hash)
  logDebug(expiration.String())
  logDebug(time.Now().String())
  if ! burn_after_reading {
    if config.InProduction {
      loc, _ := time.LoadLocation("")
      ret = strconv.FormatFloat(expiration.Sub(time.Now().In(loc)).Hours(), 'f', 2, 64)
    } else {
      ret = strconv.FormatFloat(expiration.Sub(time.Now()).Hours(), 'f', 2, 64)
    }
    ret = ret + " hours left to read paste."
  } else {
    if burn_after_reading {
      ret = "Paste set to burn after reading."
    }
  }
  return ret
}





html/assets/view.js
/**
 * Get a key from the page URI.
 * @returns {Uint8Array} - the key from the URI
*/
function getKeyFromUri() {
  var ret = "";
  var hex_key = window.location.hash.slice(1);
  ret = aesjs.utils.hex.toBytes(hex_key);
  console.log(ret);
  return ret;
}

/**
 * Decrypt cypertext using a given key.
 * @param {string} cyphertext - the cyphertext to decrypt
 * @param {Uint8Array} key - the key with which to decrypt the message
 * @returns {string} - cleartext for the given cyphertext
*/
function decryptMsg(cyphertext, key) {
  var ret = "";
  var encrypted_bytes = aesjs.utils.hex.toBytes(cyphertext);
  var aes_ctr = new aesjs.ModeOfOperation.ctr(key, new aesjs.Counter(5));
  var cleartext_bytes = aes_ctr.decrypt(encrypted_bytes);
  ret = aesjs.utils.utf8.fromBytes(cleartext_bytes);
  return ret;
}

$(document).ready(function() {
  var key = getKeyFromUri();
  var msg_container = document.getElementById("cyphertext");
  var cyphertext = msg_container.innerHTML;
  var cleartext = decryptMsg(cyphertext, key);
  cleartext = cleartext.replace(/\n/g, "<br/>");
  msg_container.innerHTML = cleartext;
});


html/assets/paste.js
/**
 * Generate a new cryptographically random key.
 * @returns {Uint8Array} - a random key
*/
function getNewKey() {
  var ret = new Uint8Array(256 / 8);
  window.crypto.getRandomValues(ret);
  return ret;
}

/**
 * Encrypt a message with a given key
 * @param {string} cleartext - the text to encrypt
 * @param {Uint8Array} key - the key to use for encryption
 * @returns {string} - the cyphertext
*/
function encryptMsg(cleartext, key) {
  var ret = "";
  var text_bytes = aesjs.utils.utf8.toBytes(cleartext);
  var aes_ctr = new aesjs.ModeOfOperation.ctr(key, new aesjs.Counter(5));
  var encrypted_bytes = aes_ctr.encrypt(text_bytes);
  ret = aesjs.utils.hex.fromBytes(encrypted_bytes);
  return ret;
}

/**
 * Encrypt a message to be sent to the server
 * @returns {Boolean} - true
*/
function encryptPaste() {
  var ret = true;
  var key = getNewKey();
  var cleartext = document.paste_form.paste_value.value;
  var cyphertext = encryptMsg(cleartext, key);
  var hex_key = aesjs.utils.hex.fromBytes(key);
  document.paste_form.paste_key.value = hex_key;
  document.paste_form.paste_value.value = cyphertext;
  return true;
}


html/assets/script.js
// An example 128-bit key (16 bytes * 8 bits/byte = 128 bits)
var key = [ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16 ];

// Convert text to bytes
var text = 'Text may be any length you wish, no padding is required.';
var textBytes = aesjs.utils.utf8.toBytes(text);

// The counter is optional, and if omitted will begin at 1
var aesCtr = new aesjs.ModeOfOperation.ctr(key, new aesjs.Counter(5));
var encryptedBytes = aesCtr.encrypt(textBytes);

// To print or store the binary data, you may convert it to hex
var encryptedHex = aesjs.utils.hex.fromBytes(encryptedBytes);
console.log(encryptedHex);
// "a338eda3874ed884b6199150d36f49988c90f5c47fe7792b0cf8c7f77eeffd87
//  ea145b73e82aefcf2076f881c88879e4e25b1d7b24ba2788"

// When ready to decrypt the hex string, convert it back to bytes
var encryptedBytes = aesjs.utils.hex.toBytes(encryptedHex);

// The counter mode of operation maintains internal state, so to
// decrypt a new instance must be instantiated.
var aesCtr = new aesjs.ModeOfOperation.ctr(key, new aesjs.Counter(5));
var decryptedBytes = aesCtr.decrypt(encryptedBytes);

// Convert our bytes back into text
var decryptedText = aesjs.utils.utf8.fromBytes(decryptedBytes);
console.log(decryptedText);

/**
  Function to handle the encryption 
  @param pasteval [string] the values to post
  @return [string] the value of the encrypted string
*/
function encryptPasteVal(pasteval, key) {
    var ret="";

    // Convert text to bytes
    var text = 'Text may be any length you wish, no padding is required.';
    var textBytes = aesjs.utils.utf8.toBytes(pasteval);
    
    // The counter is optional, and if omitted will begin at 1
    var aesCtr = new aesjs.ModeOfOperation.ctr(key, new aesjs.Counter(5));
    var encryptedBytes = aesCtr.encrypt(textBytes);
    
    // To print or store the binary data, you may convert it to hex
    ret = aesjs.utils.hex.fromBytes(encryptedBytes);
    
    return ret;
}

var global_key=[];

window.addEvent('domready', function() {
    $('online_form').addEventListener('submit', function() { 
        var key = [1,2,3,4,5,6,7,8,9,1,2,3,4,5,6,7];
        var pasteval=$('pasteval').value;
        $('pasteval').value=encryptPasteVal(pasteval, key);
    });
});

            


server.go
// Cryptbin
// (C) Copyright George Saussy 2018
// version 0.1
package main

import (
  "database/sql"
  "flag"
  "fmt"
  "hash/fnv"
  "html/template"
  "net/http"
  "strconv"
  "strings"
  "time"
  _ "github.com/mattn/go-sqlite3"
)

// Structs

// ServerConfig contains the configuration of the server. 
// Should contain all of the global information for the server.
type ServerConfig struct {
  InProduction bool
  Debug bool
  Password string
  Username string
  Database *sql.DB
}

// Message struct is designed to hold a stored message. 
// The struct can also be passed to the view template to render
// the message.
type Message struct {
  Key string
  Text string
}

// Functions

// sendToLog sends information to console, or a log file if set
func sendToLog(s string, config * ServerConfig) {
  fmt.Printf("%s\n", s)
}

// logInfo prints information to console, or a log file if set
func logInfo(s string, config * ServerConfig) {
  sendToLog("INFO: " + s, config)
}

// logErrors prints error information to the log
func logError(s string, config * ServerConfig) {
  sendToLog("ERROR: " + s, config)
}

// logDebug prints debugging information to the log
func logDebug(s string, config * ServerConfig) {
  if config.Debug {
    sendToLog("DEBUG: "+ s, config)
  }
}

func checkError(err error) {
  if err != nil {
    panic(err)
  }
}

// checkAndInitDataBase checks if a database is present.
// If no database present, it initializes one.
func checkAndInitDataBase(config *ServerConfig) {
  if config.Database != nil {
    logDebug("Database initialized!", config)
  }
}

// pasteHandler responds to the `/paste/` URI and sends the
// paste home page located at `html/paste.html`.
func pasteHandler(w http.ResponseWriter, r *http.Request, config * ServerConfig) {
  logInfo("Rendering paste.", config)
  t, _ := template.ParseFiles("html/paste.html")
  t.Execute(w, nil)
}

// hashString wraps the FNV hash function to hash strings
func hashString(s string) string {
  ret := ""
  h := fnv.New64a()
  h.Write([]byte(s))
  ret = strconv.FormatUint(h.Sum64(), 16) // XXX check conversion for correctness
  return ret
}

// newHandler responds to POST requests posting a new form.
// A correctly formated POST is then redirected to the view 
// the submited content.
func newHandler(w http.ResponseWriter, r *http.Request, config * ServerConfig) {
  logInfo("Computing new.", config)
  created_at := time.Now()
  paste_value := r.FormValue("paste_value")
  expiration_text := r.FormValue("expiration")
  expiration := time.Now()
  burn_after_reading := false
  switch expiration_text {
  case "burn_after_reading":
    burn_after_reading = true
    expiration = expiration.Add(time.Hour*12)
  case "12_hr":
    expiration = expiration.Add(time.Hour*12)
  case "24_hr":
    expiration = expiration.Add(time.Hour*24)
  case "5_days":
    expiration = expiration.Add(time.Hour*5*24)
  default:
    expiration = expiration.Add(time.Hour*12) // Default to 12 hr 
  }
  paste_hash := hashString(paste_value)
  logDebug("Value: " + paste_value, config)
  logDebug("Expiration: " + expiration_text, config)
  logDebug("Hash: " + paste_hash, config)
  stmt, _ := config.Database.Prepare("INSERT INTO paste(created_at, paste_value, expiration, burn_after_reading, paste_hash) values(?, ?, ?, ?, ?)")
  stmt.Exec(created_at, paste_value, expiration, burn_after_reading, paste_hash)
  http.Redirect(w, r, "/view/"+paste_hash, 301)
}

// aboutHandler responds to the `/about/` URI and sends the
// about page for Cryptbin located at `html/about.html`.
func aboutHandler(w http.ResponseWriter, r *http.Request, config *ServerConfig) {
  logInfo("Rendering about.", config)
  t, _ := template.ParseFiles("html/about.html")
  t.Execute(w, nil)
}

// checkValidHash checks if a string is a hexidecimal string
func checkValidHash(paste_hash string) bool {
  ret := true
  valid_chars := "0123456789abcdefABCDEF"
  for ind := 0; ind < len(paste_hash) && ret; ind++ {
    if ! strings.Contains(valid_chars, paste_hash[ind:ind+1]) {
      ret = false
    }
  }
  return ret
}

// viewHandler responds to the `/view/` URI and sends the 
// view page template located at `html/view.html` using
// information given in te URI.
func viewHandler(w http.ResponseWriter, r *http.Request, config *ServerConfig) {
  logInfo("Rendering view.", config)
  uri := r.URL.Path
  paste_hash := uri[len("/view/"):]
  if checkValidHash(paste_hash) {
    rows, _ := config.Database.Query("SELECT paste_value FROM paste WHERE paste_hash='" + paste_hash + "'")
    paste_value := "[tktk]"
    for rows.Next() {
      rows.Scan(&paste_value)
    }
    rows.Close()
    logDebug("URI: " + uri, config)
    logDebug("hash: " + paste_hash, config)
    logDebug("paste: " + paste_value, config)
    dat := &Message{Key: paste_hash, Text: paste_value}
    t, _ := template.ParseFiles("html/view.html")
    t.Execute(w, dat)
  } else { // in case hash is invalid, redirect home
    // FIXME implement a 404 page
    pasteHandler(w, r, config)
  }
}

// makeHandler wraps the HandlerFuncs to pass the server configuration.
func makeHandler(fn func(http.ResponseWriter, *http.Request, *ServerConfig), config *ServerConfig) http.HandlerFunc {
  return func(w http.ResponseWriter, r *http.Request) {
    fn(w, r, config)
  }
}

// main function is just the main function of this code base.
// As of right now, all that main does is 
// parse the command line arguments,
// initialize global variables, and 
// initialize an http server accordingly. 
func main() {
  // Parse flags
  inProduction :=  flag.Bool("prod", false, "Set as true if the server is running in production.")
  debug := flag.Bool("debug", false, "Set to true to print debugging information to log.")
  password := flag.String("pwd", "", "Set the Postgres database password")
  username := flag.String("uname", "", "Set the Postgres username.")
  flag.Parse()
  config := &ServerConfig{InProduction: *inProduction, Debug: *debug, Password: *password, Username: *username, Database: nil}
  // Init database if need be  
  db, err := sql.Open("sqlite3", "db/cryptbin.db")
  if err != nil {
    logError("Database is busted", config)
  } else {
    config.Database = db
  }
  checkAndInitDataBase(config)
  // Set up http server
  http.HandleFunc("/paste/", makeHandler(pasteHandler, config)) // Route `paste` to ''
  http.HandleFunc("/new/", makeHandler(newHandler, config)) // Route `new` to ''
  http.HandleFunc("/view/", makeHandler(viewHandler, config)) // Route `view` to ''
  http.HandleFunc("/about/", makeHandler(aboutHandler, config)) // Route `about` to ''
  http.HandleFunc("/", makeHandler(pasteHandler, config)) // Route root to `paste`
  if * inProduction {
    fmt.Printf("Running Cryptbin in production mode\n")
    fmt.Printf("Go to localhost:8080\n")
    http.ListenAndServe(":8080",nil)
  } else {
    fmt.Printf("Running Cryptbin in developer mode\n")
    fmt.Printf("Go to localhost:8000\n")
    http.ListenAndServe(":8000", nil)
  }
}
